File: src/main/java/io/github/erp/service/TaxRuleService.java
package io.github.erp.service;

/*-
 * Erp System - Mark X No 8 (Jehoiada Series) Server ver 1.8.0
 * Copyright © 2021 - 2024 Edwin Njeru and the ERP System Contributors (mailnjeru@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import io.github.erp.service.dto.TaxRuleDTO;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service Interface for managing {@link io.github.erp.domain.TaxRule}.
 */
public interface TaxRuleService {
    /**
     * Save a taxRule.
     *
     * @param taxRuleDTO the entity to save.
     * @return the persisted entity.
     */
    TaxRuleDTO save(TaxRuleDTO taxRuleDTO);

    /**
     * Partially updates a taxRule.
     *
     * @param taxRuleDTO the entity to update partially.
     * @return the persisted entity.
     */
    Optional<TaxRuleDTO> partialUpdate(TaxRuleDTO taxRuleDTO);

    /**
     * Get all the taxRules.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<TaxRuleDTO> findAll(Pageable pageable);

    /**
     * Get all the taxRules with eager load of many-to-many relationships.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<TaxRuleDTO> findAllWithEagerRelationships(Pageable pageable);

    /**
     * Get the "id" taxRule.
     *
     * @param id the id of the entity.
     * @return the entity.
     */
    Optional<TaxRuleDTO> findOne(Long id);

    /**
     * Delete the "id" taxRule.
     *
     * @param id the id of the entity.
     */
    void delete(Long id);

    /**
     * Search for the taxRule corresponding to the query.
     *
     * @param query the query of the search.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<TaxRuleDTO> search(String query, Pageable pageable);
}

File: erp-client/src/main/webapp/app/entities/tax/tax-rule/service/tax-rule.service.ts
///
/// Erp System - Mark X No 10 (Jehoiada Series) Client 1.7.8
/// Copyright © 2021 - 2024 Edwin Njeru (mailnjeru@gmail.com)
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program. If not, see <http://www.gnu.org/licenses/>.
///

import { Injectable } from '@angular/core';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';

import { isPresent } from 'app/core/util/operators';
import { ApplicationConfigService } from 'app/core/config/application-config.service';
import { createRequestOption } from 'app/core/request/request-util';
import { SearchWithPagination } from 'app/core/request/request.model';
import { ITaxRule, getTaxRuleIdentifier } from '../tax-rule.model';

export type EntityResponseType = HttpResponse<ITaxRule>;
export type EntityArrayResponseType = HttpResponse<ITaxRule[]>;

@Injectable({ providedIn: 'root' })
export class TaxRuleService {
  protected resourceUrl = this.applicationConfigService.getEndpointFor('api/tax-rules');
  protected resourceSearchUrl = this.applicationConfigService.getEndpointFor('api/_search/tax-rules');

  constructor(protected http: HttpClient, protected applicationConfigService: ApplicationConfigService) {}

  create(taxRule: ITaxRule): Observable<EntityResponseType> {
    return this.http.post<ITaxRule>(this.resourceUrl, taxRule, { observe: 'response' });
  }

  update(taxRule: ITaxRule): Observable<EntityResponseType> {
    return this.http.put<ITaxRule>(`${this.resourceUrl}/${getTaxRuleIdentifier(taxRule) as number}`, taxRule, { observe: 'response' });
  }

  partialUpdate(taxRule: ITaxRule): Observable<EntityResponseType> {
    return this.http.patch<ITaxRule>(`${this.resourceUrl}/${getTaxRuleIdentifier(taxRule) as number}`, taxRule, { observe: 'response' });
  }

  find(id: number): Observable<EntityResponseType> {
    return this.http.get<ITaxRule>(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  query(req?: any): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<ITaxRule[]>(this.resourceUrl, { params: options, observe: 'response' });
  }

  delete(id: number): Observable<HttpResponse<{}>> {
    return this.http.delete(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  search(req: SearchWithPagination): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<ITaxRule[]>(this.resourceSearchUrl, { params: options, observe: 'response' });
  }

  addTaxRuleToCollectionIfMissing(taxRuleCollection: ITaxRule[], ...taxRulesToCheck: (ITaxRule | null | undefined)[]): ITaxRule[] {
    const taxRules: ITaxRule[] = taxRulesToCheck.filter(isPresent);
    if (taxRules.length > 0) {
      const taxRuleCollectionIdentifiers = taxRuleCollection.map(taxRuleItem => getTaxRuleIdentifier(taxRuleItem)!);
      const taxRulesToAdd = taxRules.filter(taxRuleItem => {
        const taxRuleIdentifier = getTaxRuleIdentifier(taxRuleItem);
        if (taxRuleIdentifier == null || taxRuleCollectionIdentifiers.includes(taxRuleIdentifier)) {
          return false;
        }
        taxRuleCollectionIdentifiers.push(taxRuleIdentifier);
        return true;
      });
      return [...taxRulesToAdd, ...taxRuleCollection];
    }
    return taxRuleCollection;
  }
}
